### PR TITLE
Add span on finish callbacks

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
## Description

With these callbacks we could eventually remove the need to specify a
tracer or context instance to ``Span``.

## Checklist
- [x] Added to the correct milestone.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Library documentation is updated.
- [x] ~[Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).~

### CODE REVIEW

**Review by P403n1x87 — APPROVED** [review]
👍

**Comment by majorgreys** [comment]
How do distinguish between when to use `callback` vs. `hook`? We have tracer hooks for specific actions. My thoughts here would be that this `on_finish_callback` is more of a callback than a hook since it is applied at the end of the lifecycle for a span. However I feel like callback is meaningful in an asynchronous function call whereas this case are functions that are called when a synchronous function is called, either by the user or by the library itself. Therefore it would seem a better case for keeping the `hook` naming.

**Review by P403n1x87 — APPROVED** [review]

**Comment by Kyle-Verhoog** [comment]
@majorgreys we don't actually use the word "hook" in our API for tracer (or the documentation for it) or the word "callback" in the public API of the span here (unless I'm missing something). These terms are only used internally by us so I don't think it's an issue from a user point of view.

If you think it'll be an issue for us internally down the line then I can make the change but I think it would be pretty clear to a developer working on ddtrace that the `_on_finish_callbacks` are synchronous.

**Review by Yun-Kim — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Comment by majorgreys** [comment]
> These terms are only used internally by us so I don't think it's an issue from a user point of view.

@Kyle-Verhoog Right, I was confusing this more library/sdk/internal cocnern with a public API for a user of the tracer.

File: ddtrace/span.py [hunk 1]
```diff
@@ -206,6 +211,9 @@ def finish(self, finish_time=None):
             if self.tracer and trace and sampled:
                 self.tracer.write(trace)
 
+        for cb in self._on_finish_callbacks:
+            cb(self)
```
Comment (jd) [inline ddtrace/span.py:215]: No error handling there?

Comment (Kyle-Verhoog) [reply ddtrace/span.py:215]: I'm not sure what the error handling would accomplish. If the callback raised then I think that'd be the responsibility of the own of a `Span` to handle.

**Pull request merged**